spec/reek/tree_walker_spec.rb

```diff
@@ -1,237 +1,238 @@
+      end
+
+      it 'counts 3 statements in a rescue' do
+        code = <<-EOS
+          def one()
+            begin
+              callee(); callee(); callee()
+            rescue
+              callee(); callee(); callee()
+            end
           end
-        end
-                              ')
-      expect(method.num_statements).to eq(3)
-    end
-
-    it 'counts 1 statement in a case else' do
-      source = 'def one() case fred; when "hi"; callee(); else; callee(); end; end'
-      expect(process_method(source).num_statements).to eq(2)
-    end
-
-    it 'counts 3 statements in a case else' do
-      method = process_method('
-        def one()
-          case fred
-          when "hi" then callee(); callee(); callee()
-          else           callee(); callee(); callee()
+        EOS
+        expect(number_of_statements_for(code)).to eq(6)
+      end
+
+      it 'counts 3 statements in a when' do
+        code = <<-EOS
+          def one()
+            case fred
+            when "hi" then callee(); callee()
+            when "lo" then callee()
+            end
           end
-        end
-                              ')
-      expect(method.num_statements).to eq(6)
-    end
-
-    it 'does not count empty case' do
-      method = process_method('def one() case fred; when "hi"; ; when "lo"; ; end; end')
-      expect(method.num_statements).to eq(0)
-    end
-
-    it 'does not count empty case else' do
-      method = process_method('def one() case fred; when "hi"; ; else; ; end; end')
-      expect(method.num_statements).to eq(0)
+        EOS
+        expect(number_of_statements_for(code)).to eq(3)
+      end
+
+      it 'counts 3 statements in a case else' do
+        code = <<-EOS
+          def one()
+            case fred
+            when "hi" then callee(); callee(); callee()
+            else           callee(); callee(); callee()
+            end
+          end
+        EOS
+        expect(number_of_statements_for(code)).to eq(6)
+      end
+
+      it 'does not count empty case' do
+        code = 'def one() case fred; when "hi"; ; when "lo"; ; end; end'
+        expect(number_of_statements_for(code)).to eq(0)
+      end
+
+      it 'does not count empty case else' do
+        code = 'def one() case fred; when "hi"; ; else; ; end; end'
+        expect(number_of_statements_for(code)).to eq(0)
+      end
+
+      it 'counts 4 statements in an iterator' do
+        code = 'def one() fred.each do; callee(); callee(); callee(); end; end'
+        expect(number_of_statements_for(code)).to eq(4)
+      end
+
+      it 'counts 1 statement in a singleton method' do
+        code = 'def self.foo; callee(); end'
+        expect(number_of_statements_for(code)).to eq(1)
+      end
     end
+  end
 
-    it 'counts 2 statement in an iterator' do
-      method = process_method('def one() fred.each do; callee(); end; end')
-      expect(method.num_statements).to eq(2)
-    end
+  describe '#walk' do
```

Comment:
Agreed on revisiting this later. See also the SMELL and TODO comments at the top of TreeWalker. Ideally, we can expose `#process` as a public method and have TreeWalker just produce context trees that we can walk elsewhere.